uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -8,68 +8,26 @@
+//! 1. Call the scaffolding function to get a [RustFutureHandle]
+//! 2a. In a loop:
+//!   - Call [rust_future_poll]
+//!   - Suspend the function until the [rust_future_poll] continuation function is called
+//!   - If the continuation was function was called with [RustFuturePoll::Ready], then break
+//!     otherwise continue.
+//! 2b. If the async function is cancelled, then call [rust_future_cancel].  This causes the
+//!     continuation function to be called with [RustFuturePoll::Ready] and the [RustFuture] to
+//!     enter a cancelled state.
+//! 3. Call [rust_future_complete] to get the result of the future.
+//! 4. Call [rust_future_free] to free the future, ideally in a finally block.  This:
+//!    - Releases any resources held by the future
+//!    - Calls any continuation callbacks that have not been called yet
 //!
-//! # Anatomy of an async call
-//!
-//! Let's consider the following Rust function:
-//!
-//! ```rust,ignore
-//! #[uniffi::export]
-//! async fn hello() -> bool {
-//!     true
-//! }
-//! ```
-//!
-//! In Rust, this `async fn` syntax is strictly equivalent to a normal function that returns a
-//! `Future`:
-//!
-//! ```rust,ignore
-//! #[uniffi::export]
-//! fn hello() -> impl Future<Output = bool> { /* … */ }
-//! ```
-//!
-//! `uniffi-bindgen` will generate a scaffolding function for each exported async function:
-//!
-//! ```rust,ignore
-//! // The `hello` function, as seen from the outside. It inputs 3 extra arguments:
-//! //  - executor: used to schedule polls of the future
-//! //  - callback: invoked when the future is ready
-//! //  - callback_data: opaque pointer that's passed to the callback.  It points to any state needed to
-//! //    resume the async function.
-//! #[no_mangle]
-//! pub extern "C" fn _uniffi_hello(
-//!     // ...If the function inputted arguments, the lowered versions would go here
-//!     uniffi_executor: ForeignExecutor,
-//!     uniffi_callback: <bool as FfiConverter<crate::UniFFITag>>::FutureCallback,
-//!     uniffi_callback_data: *const (),
-//!     uniffi_call_status: &mut ::uniffi::RustCallStatus
-//! ) {
-//!     ::uniffi::call_with_output(uniffi_call_status, || {
-//!         let uniffi_rust_future = RustFuture::<_, bool, crate::UniFFITag,>::new(
-//!             future: hello(), // the future!
-//!             uniffi_executor,
-//!             uniffi_callback,
-//!             uniffi_callback_data,
-//!         );
-//!         uniffi_rust_future.wake();
-//!     })
-//! }
-//! ```
-//!
-//! Rust will continue to poll the future until it's ready, after that.  The callback will
-//! eventually be invoked with these arguments:
-//!   - callback_data
-//!   - FfiConverter::ReturnType (the type that would be returned by a sync function)
-//!   - RustCallStatus (used to signal errors/panics when executing the future)
-//! - Rust will stop polling the future, even if it's waker is invoked again.
```

Comment:
I really liked this example code as a reader, but it was really hard to keep up to date as I was making changes.  Hopefully the pseudocode version will provide close to the same benefit without being so hard to maintain.  Does that make sense to others?